src/com/amazon/ion/impl/bin/IonEncoder_1_1.java

```diff
@@ -158,4 +157,219 @@ public static int writeFloat(WriteBuffer buffer, final double value) {
+        }
+
+        // If the precision is year, month, or day, we can skip the remaining checks.
+        if (!value.getPrecision().includes(Timestamp.Precision.MINUTE)) {
+            return writeShortFormTimestampValue(buffer, value);
+        }
+
+        // Condition 2: The fractional seconds are a common precision.
+        int secondsScale = value.getDecimalSecond().scale();
+        if (secondsScale != 0 && secondsScale != 3 && secondsScale != 6 && secondsScale != 9) {
+            return writeLongFormTimestampValue(buffer, value);
+        }
+        // Condition 3: The local offset is either UTC, unknown, or falls between -14:00 to +14:00 and is divisible by 15 minutes.
+        Integer offset = value.getLocalOffset();
+        if (offset != null && (offset < -14 * 60 || offset > 14 * 60 || offset % 15 != 0)) {
+            return writeLongFormTimestampValue(buffer, value);
+        }
+        return writeShortFormTimestampValue(buffer, value);
+    }
+
+    /**
+     * Writes a short-form timestamp.
+     * Value cannot be null.
+     * If calling from outside this class, use writeTimestampValue instead.
+     */
+    private static int writeShortFormTimestampValue(WriteBuffer buffer, Timestamp value) {
+        long bits = (value.getYear() - 1970L);
+        if (value.getPrecision() == Timestamp.Precision.YEAR) {
+            buffer.writeByte(OpCodes.TIMESTAMP_YEAR_PRECISION);
+            buffer.writeFixedIntOrUInt(bits, 1);
+            return 2;
+        }
+
+        bits |= ((long) value.getMonth()) << S_TIMESTAMP_MONTH_BIT_OFFSET;
+        if (value.getPrecision() == Timestamp.Precision.MONTH) {
+            buffer.writeByte(OpCodes.TIMESTAMP_MONTH_PRECISION);
+            buffer.writeFixedIntOrUInt(bits, 2);
+            return 3;
+        }
+
+        bits |= ((long) value.getDay()) << S_TIMESTAMP_DAY_BIT_OFFSET;
+        if (value.getPrecision() == Timestamp.Precision.DAY) {
+            buffer.writeByte(OpCodes.TIMESTAMP_DAY_PRECISION);
+            buffer.writeFixedIntOrUInt(bits, 2);
+            return 3;
+        }
+
+        bits |= ((long) value.getHour()) << S_TIMESTAMP_HOUR_BIT_OFFSET;
+        bits |= ((long) value.getMinute()) << S_TIMESTAMP_MINUTE_BIT_OFFSET;
+        if (value.getLocalOffset() == null || value.getLocalOffset() == 0) {
+            if (value.getLocalOffset() != null) {
+                bits |= S_U_TIMESTAMP_UTC_FLAG;
+            }
+
+            if (value.getPrecision() == Timestamp.Precision.MINUTE) {
+                buffer.writeByte(OpCodes.TIMESTAMP_MINUTE_PRECISION);
+                buffer.writeFixedIntOrUInt(bits, 4);
+                return 5;
+            }
+
+            bits |= ((long) value.getSecond()) << S_U_TIMESTAMP_SECOND_BIT_OFFSET;
+
+            int secondsScale = value.getDecimalSecond().scale();
+            if (secondsScale != 0) {
+                long fractionalSeconds = value.getDecimalSecond().remainder(BigDecimal.ONE).movePointRight(secondsScale).longValue();
```

Comment:
I see `value.unscaledValue().longValue()` in my prototype. Maybe you can get that to work. Remember that any `BigDecimal` method that returns `BigDecimal` allocates a new one, so we want to minimize that.